app/src/main/java/pm/gnosis/heimdall/ui/transactions/BaseTransactionActivity.kt

```diff
@@ -0,0 +1,184 @@
+                        .flatMap {
+                            checkInfoAndPerform(it, { safeAddress, transaction ->
+                                viewModel.loadTransactionInfo(safeAddress, transaction)
+                            })
+                        }
+                        .observeOn(AndroidSchedulers.mainThread())
+                        .doOnNext({ it.handle(::updateInfo, ::handleInfoError) })
+            }
+
+    private val cacheTransactionInfoTransformer: ObservableTransformer<Pair<BigInteger?, Result<Transaction>>, *> =
+            ObservableTransformer { up: Observable<Pair<BigInteger?, Result<Transaction>>> ->
+                up.observeOn(AndroidSchedulers.mainThread()).doOnNext({ (safeAddress, transaction) ->
+                    transaction.handle({ currentInfo = safeAddress to it }, { currentInfo = null })
+                })
+            }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        inject()
+        setContentView(R.layout.layout_transaction_details)
+
+        setupToolbar(layout_transaction_details_toolbar, R.drawable.ic_close_24dp)
+    }
+
+    abstract fun loadTransactionDetails()
+
+    override fun onStart() {
+        super.onStart()
+        setUnknownTransactionInfo()
+        disposables += layout_transaction_details_button.clicks()
+                .flatMap {
+                    currentInfo?.let { (safeAddress, transaction) ->
+                        safeAddress?.let {
+                            viewModel.submitTransaction(safeAddress, transaction)
+                                    .subscribeOn(AndroidSchedulers.mainThread())
+                                    .observeOn(AndroidSchedulers.mainThread())
+                                    .doOnSubscribe { submittingTransaction(true) }
+                                    .doAfterTerminate { submittingTransaction(false) }
+                                    .toObservable()
+                        }
+                    } ?: Observable.empty<Result<Unit>>()
+                }.subscribeForResult({ finish() }, { showErrorSnackbar(it) })
+
+        val fragment = supportFragmentManager.findFragmentById(R.id.layout_transaction_details_transaction_container)
+        if (fragment is BaseTransactionDetailsFragment) {
+            setCurrentFragment(fragment)
+        } else {
+            loadTransactionDetails()
+        }
+    }
+
+    private fun submittingTransaction(loading: Boolean) {
+        layout_transaction_details_progress_bar.visibility = if (loading) View.VISIBLE else View.GONE
+        layout_transaction_details_button.isEnabled = !loading
+        currentFragment?.inputEnabled(!loading)
+    }
+
+    protected fun displayTransactionDetails(fragment: BaseTransactionDetailsFragment) {
+        supportFragmentManager.transaction {
+            replace(R.id.layout_transaction_details_transaction_container, fragment)
+        }
+        setCurrentFragment(fragment)
```

Comment:
switch order, just to avoid possible race conditions with registerFragmentObservables